### DISCUSSION TITLE
Performance difference between client-to-server and server-to client

### DISCUSSION THREAD

**Discussion by zhouliguo** [miguelgrinberg/python-socketio#1187] [Q&A] [answered]
I found that sending data from client to server is much slower than sending data from server to client, by about a factor of ten. And the size of the data that can be sent also varies greatly.

What is the reason？
![c](https://github.com/miguelgrinberg/python-socketio/assets/4395337/e397f8c5-921c-4360-8f64-6c8b6408ae1a)
![s](https://github.com/miguelgrinberg/python-socketio/assets/4395337/eae9c06f-1174-4bfd-b5d5-2811956e689a)

**Comment by miguelgrinberg** [comment]
My guess is that the reason is directly related to the difference in download and upload speeds of the Internet service your client uses.

**Reply by zhouliguo** [reply]
Thank you for your answer. I'm using Localhost, not the Internet, the client and server are on the same computer. Is there any other reason? python-socketio = 5.8.0

**Reply by miguelgrinberg** [reply]
You should look for answers in your own application. The Socket.IO server and client do not have anything in them that slows down traffic in one direction. You clearly aren't doing the same work in your server and your client. Maybe that's a factor?

**Reply by zhouliguo** [reply]
Thank you very much for your reply. I have checked my code several times. I put the code here. Two for transferring data from client to server,  very slow, and the data size limit is also small. Two for transferring data from server to client, very fast, and no data size limitation. They have done the same thing, but the performances are different.

```
# Client-to-Server Client
import time
import socketio
import numpy as np

sioc = socketio.Client()

@sioc.event
def connect():
    print('connected to server')
    sioc.emit(event = 'receive_frame', data='connect')

start = 0

# Send image from client to server.
@sioc.event
def send_frame():
    global start
    end = time.time() # sent and get response

    if start != 0:
        print('Sending duration', (end-start)*1000,'ms')
    image = np.random.random((512, 512, 3))
    if sioc.connected:
        start = time.time() # start sending
        sioc.emit('receive_frame', data=image.tobytes())
        

@sioc.event
def disconnect():
    quit()

if __name__ == '__main__':
    sioc.connect('http://localhost:5000')
    sioc.wait()
```

```
# Client-to-Server Server
import socketio
import eventlet

sios = socketio.Server(max_http_buffer_size = 100000000)
app = socketio.WSGIApp(socketio_app=sios)

@sios.event
def connect(sid, environ, auth):
    print('connect ', sid)

@sios.event
def receive_frame(sid, data):
    sios.emit(event = 'send_frame')
    print('received')

@sios.event
def disconnect(sid):
    print('disconnect ', sid)

if __name__ == '__main__':
    eventlet.wsgi.server(eventlet.listen(('localhost', 5000)), app)
```

```
# Server-to-Client Client
import socketio

sioc = socketio.Client()

@sioc.event
def connect():
    print('connected to server')
    sioc.emit(event = 'send_frame')

start = 0

# Send image from client to server.
@sioc.event
def receive_frame(data):
    sioc.emit(event = 'send_frame')
    print('received')
        

@sioc.event
def disconnect():
    quit()

if __name__ == '__main__':
    sioc.connect('http://localhost:5000')
    sioc.wait()
```

```
# Server-to-Client Server
import socketio
import eventlet
import time
import numpy as np

sios = socketio.Server(max_http_buffer_size = 100000000)
app = socketio.WSGIApp(socketio_app=sios)

@sios.event
def connect(sid, environ, auth):
    print('connect ', sid)

start = 0

@sios.event
def send_frame(sid):
    global start
    end = time.time() # sent and get response

    if start != 0:
        print('Sending duration', (end-start)*1000,'ms')
    image = np.random.random((2048, 2048, 3))
    start = time.time() # start sending
    sios.emit('receive_frame', data=image.tobytes())


@sios.event
def disconnect(sid):
    print('disconnect ', sid)

if __name__ == '__main__':
    eventlet.wsgi.server(eventlet.listen(('localhost', 5000)), app)
```

**Reply by miguelgrinberg** [reply]
I don't really have an answer for you, sorry. As I said, this package does not care, it just writes to the sockets without restricting anything. The server imposes a size restriction on clients as a security measure, but clearly you have discovered that setting and changed it.

Other than that, it's up to the downstream implementations and your network stack. If you are concerned about performance you should try all the supported servers. Gunicorn's eventlet worker often performs better than eventlet alone. Have you tried that? I really don't have any other ideas to explain this, sorry.

**Reply by zhouliguo** [reply, marked as answer]
I changed the 
```
@sioc.event
def send_frame()
```
to
```
sio.on('send_frame')
def send_frame()
```
in the 'Client-to-Server Client' file, the performance becomes normal. I don't know what's the difference between the two. Is this a bug?

**Reply by miguelgrinberg** [reply]
The `on()` and `event` decorators are pretty much equivalent, the only difference is how the event name is defined.

**Reply by zhouliguo** [reply]
Then I can be sure that your code has bugs, stop being so stubborn！

**Reply by miguelgrinberg** [reply]
Look at the "working code" snippet you provided me:

```
sio.on('send_frame')
def send_frame()
```

What is `sio` here? Why aren't you using a decorator syntax? How do you expect me to debug the bugs in my code when you make such a low effort in defining what the problem is? Clearly your code has bugs too, don't you think?
